src/algebra/ordered_ring_hom.lean

```diff
@@ -0,0 +1,146 @@
+instance has_coe_to_hom : has_coe (R →+*o S) (R → S) := ⟨ring_hom.to_fun ∘ to_ring_hom⟩
+instance has_coe_to_rel_hom : has_coe (R →+*o S) (((≤) : R → R → Prop) →r ((≤) : S → S → Prop)) :=
+⟨to_rel_hom⟩
+
+instance : has_coe_to_fun (R →+*o S) := ⟨_, λ f, f.to_fun⟩
+
+@[simp] lemma to_ring_hom_eq_coe {f : R →+*o S} : f.to_ring_hom = f := rfl
+@[simp] lemma to_order_iso_eq_coe {f : R →+*o S} : f.to_rel_hom = f := rfl
+@[simp] lemma coe_ring_hom_to_fun_eq_coe_fun {f : R →+*o S} : (f : R →+* S).to_fun = f := rfl
+@[simp] lemma coe_ring_hom_coe_fun_eq_coe_fun {f : R →+*o S} : ((f : R →+* S) : R → S) = f := rfl
+@[simp] lemma coe_rel_hom_to_fun_eq_to_equiv {f : R →+*o S} :
+  (f : ((≤) : R → R → Prop) →r ((≤) : S → S → Prop)).to_fun = f := rfl
+@[simp] lemma coe_rel_hom_coe_fun_eq_to_equiv {f : R →+*o S} :
+  ((f : ((≤) : R → R → Prop) →r ((≤) : S → S → Prop)) : R → S) = f := rfl
+@[simp]
+lemma coe_mul_equiv_to_fun_eq_coe_fun {f : R →+*o S} : ((f : R →+* S) : R →* S).to_fun = f := rfl
+@[simp]
+lemma coe_mul_equiv_coe_fun_eq_coe_fun {f : R →+*o S} : (((f : R →+* S) : R →* S) : R → S) = f :=
+rfl
+@[simp]
+lemma coe_add_equiv_to_fun_eq_coe_fun {f : R →+*o S} : ((f : R →+* S) : R →+ S).to_fun = f := rfl
+@[simp]
+lemma coe_add_equiv_coe_fun_eq_coe_fun {f : R →+*o S} : (((f : R →+* S) : R →+ S) : R → S) = f :=
+rfl
+
+/-- An ordered ring homomorphism preserves multiplication. -/
+@[simp] lemma map_mul (e : R →+*o S) (x y : R) : e (x * y) = e x * e y := e.map_mul' x y
+
+/-- An ordered ring homomorphism preserves addition. -/
+@[simp] lemma map_add (e : R →+*o S) (x y : R) : e (x + y) = e x + e y := e.map_add' x y
+
+/-- An ordered ring homomorphism preserves ordering. -/
+@[simp] lemma map_le (e : R →+*o S) {x y : R} (h : x ≤ y) : e x ≤ e y := e.map_rel' h
+
+protected lemma congr_arg {f : R →+*o S} : Π {x x' : R}, x = x' → f x = f x'
+| _ _ rfl := rfl
+
+protected lemma congr_fun {f g : R →+*o S} (h : f = g) (x : R) : f x = g x := h ▸ rfl
+
+/-- Two ordered ring homomorphisms agree if they are defined by the
+  same underlying function. -/
+@[ext] lemma ext {f g : R →+*o S} (h : ∀ x, f x = g x) : f = g :=
+begin
+  cases f, cases g, congr,
+  apply ring_hom.ext,
+  intro x,
+  convert h x,
+end
+
+lemma ext_iff {f g : R →+*o S} : f = g ↔ ∀ x, f x = g x := ⟨λ h x, h ▸ rfl, ext⟩
+
+@[norm_cast] lemma coe_ring_equiv (f : R →+*o S) (a : R) : (f : R →+* S) a = f a := rfl
+
+@[norm_cast] lemma coe_mul_equiv (f : R →+*o S) (a : R) : (f : R →* S) a = f a := rfl
+
+@[norm_cast] lemma coe_add_equiv (f : R →+*o S) (a : R) : (f : R →+ S) a = f a := rfl
+
+variable (R)
+
+/-- Identity map is an ordered ring isomorphism. -/
+protected def id : R →+*o R :=
+{ ..ring_hom.id _,
+  ..rel_hom.id _, }
+
+variable {R}
+@[simp] lemma id_apply (x : R) : ordered_ring_hom.id R x = x := rfl
+
+@[simp] lemma coe_add_hom_id : (ordered_ring_hom.id R : R →+ R) = add_monoid_hom.id R := rfl
+@[simp] lemma coe_mul_hom_id : (ordered_ring_hom.id R : R →* R) = monoid_hom.id R := rfl
```

Comment:
I think these two lemmas belong in `algebra.ring.basic`, and should state:
```suggestion
@[simp] lemma coe_add_monoid_hom_id : (ring_hom.id R : R →+ R) = add_monoid_hom.id R := rfl
@[simp] lemma coe_monoid_hom_id : (ring_hom.id R : R →* R) = monoid_hom.id R := rfl
```
That should be enough to make these provable as a result of `coe_ring_hom_id` below.